vimiv/imutils/metadata.py

```diff
@@ -0,0 +1,284 @@
+
+        If no value is found for a certain key, do not include the key in the output.
+
+        Args:
+            _keys: Keys of metadata to query the image for.
+
+        Returns:
+            Dictionary with retrieved metadata.
+        """
+
+    @abstractmethod
+    def get_keys(self) -> Iterable[str]:
+        """Get the keys for all metadata values available for the current image."""
+
+    def copy_metadata(self, _dest: str, _reset_orientation: bool = True) -> bool:
+        """Copy metadata from the current image to dest image.
+
+        Args:
+            _dest: Path to write the metadata to.
+            _reset_orientation: If true, reset the exif orientation tag to normal.
+
+        Returns:
+            Flag indicating if copy was successful.
+        """
+        raise NotImplementedError
+
+    def get_date_time(self) -> str:
+        """Get creation date and time of the current image as formatted string."""
+        raise NotImplementedError
+
+
+# Stores all registered metadata implementations.
+_registry: List[Type[MetadataPlugin]] = []
+
+
+def has_metadata_support() -> bool:
+    """Indicate if `MetadataHandler` has `get_metadata()` and `get_keys()` capabilities.
+
+    Returns:
+        True iff at least one metadata plugins has been registered.
+    """
+    return bool(_registry)
+
+
+class MetadataHandler:
+    """Handle metadata related functionalities of images.
+
+    Attributes:
+        _path: Path to current image.
+    """
+
+    def __init__(self, path: str):
+        self._path = path
+
+    @property
+    def has_copy_metadata(self) -> bool:
+        """True iff `MetadataHandler` has an implementation for `copy_metadata`."""
+        return any(
+            list(
+                map(
+                    lambda e: e.copy_metadata != MetadataPlugin.copy_metadata, _registry
+                )
+            )
+        )
+
+    @property
+    def has_get_date_time(self) -> bool:
+        """True iff `MetadataHandler` has an implementation for `get_date_time`."""
+        return any(
+            list(
+                map(
+                    lambda e: e.get_date_time != MetadataPlugin.get_date_time, _registry
+                )
+            )
+        )
+
+    def get_metadata(self, keys: Sequence[str]) -> MetadataDictT:
+        """Get value of all desired keys from the current image.
+
+        Use all registered metadata implementations to extract the metadata from the
+        current image. The output of all methods is combined.
+
+        Args:
+            keys: Keys of metadata to query the image for.
+
+        Returns:
+            Dictionary with retrieved metadata.
+        """
+        out: MetadataDictT = {}
+
+        for backend in _registry:
+            with contextlib.suppress(NotImplementedError):
+                # TODO: from 3.9 on use: c = a | b
+                out = {**backend(self._path).get_metadata(keys), **out}
+
+        # TODO: ensure keys are returned in the right order
```

Comment:
What would the "right" order be? Ideally they don't overlap, but if they do, just the user-defined order, no?

Reply:
I meant that the metadata is displayed in the widget in the order the keys are specified in the config. I have implemented that in dc830215b2aebf76d5666cf758c75f263dcabdc8

Reply:
Makes sense, thanks :+1: